src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs

```diff
@@ -149,6 +229,9 @@ public void DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEve
         /// <inheritdoc/>
         public void Abort()
         {
+            // Make sure the proxy operation manager is initialized before anything.
+            this.proxyOperationManagerInitializedEvent.WaitOne();
```

Comment:
I'm not sure this (i.e. not issuing a discovery/run request) would cause a literal deadlock, rather indefinite waiting, which shouldn't be that bad given that calling `Abort` with nothing to actually abort is not really encouraged.

The need for introducing synchronization here is owed to the fact that the `ProxyOperationManager` is not necessarily initialized when the object is created. True, that's the case when we don't use test sessions, but with test sessions the `ProxyOperationManager` is only initialized when a discovery/run request is issued, making an abort before even running discovery to crash. Of course, we can do some null checking and have `Abort` do nothing if the proxy is not initialized yet, but in this way at least there's visible consequences to this behavior and maybe that can direct the caller to fix it rather than just ignore the abort request on TP side.

Reply:
that is somewhat concerning. I could imagine writing code:

```
using (var _ = ct.Register(() => tp.Abort()))
{
  ct.ThrowIfCancellationRequested();
  tp.DiscoverTests(...);
}
```

where if cancellation occurs I might not trigger a discover/tests call, but will have registered an abort handler.

in general that's the result of the existing sync API because Discover/Run blocks the caller's thread. as a result we register a cancellation token handler before calling Discover/Run methods. this means that it would be possible for TP to receive an Abort call before a Discover/Run calls

if TP would take a cancellation token or send an event that a Discovery/Run became abortable, we could ensure to never call Abort before Discover/Run

Reply:
I think it's best to keep it simple and replace the `ManualResetEvent` with null checks then.